src/controller.rs

```diff
@@ -185,26 +185,124 @@ fn apply_mappings(
         debug!(%dbg, "before");
 
         debug!(?subtree, ?mapping.to_field_path, "to field path");
-        if mapping.to_field_path.starts_with("metadata.") {
-            // DynamicObject's metadata is not a serde_json::value::Value,
-            // but we can convert to/from that and use the same code to update
-            // it as we do for spec/status in .data
-            let mut metadata = serde_json::value::to_value(template.metadata.clone())?;
-            set_field_path(
-                &mut metadata,
-                &mapping.to_field_path.strip_prefix("metadata.").unwrap(),
-                subtree.clone(),
-            )?;
-            template.metadata = serde_json::value::from_value(metadata)?;
-        } else {
-            set_field_path(&mut template.data, &mapping.to_field_path, subtree.clone())?;
+        match mapping {
+            Mapping {
+                from_field_path: None,
+                to_field_path: None,
+            } => {
+                // user must specify either from, to or both, but not neither.
+                // leave the mapping array empty if that's what they want.
+                return Err(Error::MappingEmpty);
+            }
+            Mapping {
+                from_field_path: Some(_),
+                to_field_path: None,
+            } => {
```

Comment:
this block is the bulk of the change to be reviewed, and the supporting functions it calls